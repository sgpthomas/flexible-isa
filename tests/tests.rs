use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

// #[test]
// fn parse_tests() {
//     todo!()
// }

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[&((*i) as usize)].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}

#[test]
fn add_isa() -> anyhow::Result<()> {
    test_single_isa("add")
}

#[test]
fn average_pool_isa() -> anyhow::Result<()> {
    test_single_isa("average_pool")
}

#[test]
fn blur3x3_isa() -> anyhow::Result<()> {
    test_single_isa("blur3x3")
}

#[test]
fn camera_pipe_isa() -> anyhow::Result<()> {
    test_single_isa("camera_pipe")
}

#[test]
fn conv3x3a16_isa() -> anyhow::Result<()> {
    test_single_isa("conv3x3a16")
}

#[test]
fn conv3x3a32_isa() -> anyhow::Result<()> {
    test_single_isa("conv3x3a32")
}

#[test]
fn conv_nn_isa() -> anyhow::Result<()> {
    test_single_isa("conv_nn")
}

#[test]
fn depthwise_conv_isa() -> anyhow::Result<()> {
    test_single_isa("depthwise_conv")
}

#[test]
fn dilate3x3_isa() -> anyhow::Result<()> {
    test_single_isa("dilate3x3")
}

#[test]
fn gaussian3x3_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian3x3")
}

#[test]
fn gaussian5x5_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian5x5")
}

#[test]
fn gaussian7x7_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian7x7")
}

#[test]
fn l2norm_isa() -> anyhow::Result<()> {
    test_single_isa("l2norm")
}

#[test]
fn matmul_isa() -> anyhow::Result<()> {
    test_single_isa("matmul")
}

#[test]
fn max_pool_isa() -> anyhow::Result<()> {
    test_single_isa("max_pool")
}

#[test]
fn mean_isa() -> anyhow::Result<()> {
    test_single_isa("mean")
}

#[test]
fn median3x3_isa() -> anyhow::Result<()> {
    test_single_isa("median3x3")
}

#[test]
fn mul_isa() -> anyhow::Result<()> {
    test_single_isa("mul")
}

#[test]
fn sobel3x3_isa() -> anyhow::Result<()> {
    test_single_isa("sobel3x3")
}

#[test]
fn softmax_isa() -> anyhow::Result<()> {
    test_single_isa("softmax")
}
