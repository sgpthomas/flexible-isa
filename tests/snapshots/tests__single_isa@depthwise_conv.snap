---
source: tests/tests.rs
description: generate isa for depthwise_conv
---
[
    (
        46,
        1,
        "(+ (cast<int32x128> (/ (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> ?#3))))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?x1))) (cast<int32> ?#5))) (cast<int32> ?x2))))))",
    ),
    (
        123,
        1,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        127,
        1,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 384) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        153,
        1,
        "(- (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        173,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        189,
        1,
        "(< (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        201,
        1,
        "(if (cast<uint8x128> (get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))) (cast<bool> ?x2))",
    ),
    (
        206,
        1,
        "(if ?x0 (cast<bool> ?x1))",
    ),
    (
        208,
        1,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        209,
        1,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        6,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> ?#3))) (cast<int32> ?x0))",
    ),
    (
        7,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> ?#4))) (cast<int32> ?x1))",
    ),
    (
        10,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?#2))) (cast<int32> ?#3))) (cast<int32> (- (cast<int32> ?#4) (cast<int32> ?x0))))",
    ),
    (
        25,
        2,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> 0) (cast<int32> ?x1))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> ?x3) (cast<int32> ?x4))) (cast<int32> ?x5))) (cast<int32> ?x6))))",
    ),
    (
        43,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> ?x1))))",
    ),
    (
        125,
        2,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        159,
        2,
        "(- (cast<int32> (call[max] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?#18) (cast<int32> ?x0) (cast<int32> (call[select] (cast<bool> ?x1) (cast<int32> 0) (cast<int32> ?x2))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x3) (cast<int32> ?x4) (cast<int32> ?x5))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x8))) (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?#19) (cast<int32> ?x9) (cast<int32> (call[select] (cast<bool> ?x1) (cast<int32> ?x2) (cast<int32> 0))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x3) (cast<int32> ?x5) (cast<int32> ?x4))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x10))))",
    ),
    (
        166,
        2,
        "(* (cast<int32> (+ (cast<int32> (- (cast<int32> ?#2) (cast<int32> ?x0))) (cast<int32> ?x1))) (cast<int32> ?#3))",
    ),
    (
        172,
        2,
        "(* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> ?x4))",
    ),
    (
        179,
        2,
        "(* (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        183,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> 2))",
    ),
    (
        184,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        190,
        2,
        "(<= (cast<int16x64> (call[int16x64] (cast<int32x64> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 64))))) (cast<int16x64> (call[x64] (cast<int16> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#1) (cast<int32> ?x0))))))))",
    ),
    (
        199,
        2,
        "(|| (cast<bool> (<= (cast<int32> 8) (cast<int32> ?x0))) (cast<bool> (== (cast<int32> (\"%\" (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> 0))))",
    ),
    (
        203,
        2,
        "(if (cast<uint8x128> (get ?#1 (cast<int32x128> ?x0))) (cast<bool> (<= (cast<int16x128> (call[int16x128] (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#2) (cast<int32> 1))))))))))",
    ),
    (
        204,
        2,
        "(if (cast<uint8x64> (get ?#3 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 64))))) (cast<bool> (<= (cast<int8x64> (call[int8x64] (cast<int32x64> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 64))))) (cast<int8x64> (call[x64] (cast<int8> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#4) (cast<int32> ?x2))))))))))",
    ),
    (
        229,
        2,
        "(get ?#1 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> 64))) (cast<int32> 1) (cast<int32> 64))))",
    ),
    (
        231,
        2,
        "(get ?#2 (cast<int32x128> (+ (cast<int32x128> (/ (cast<int32x128> (call[ramp] (cast<int32> ?#3) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> ?#4))))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?x0))) (cast<int32> ?#6))) (cast<int32> ?x1))))))))",
    ),
    (
        232,
        2,
        "(get ?#2 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?x0))) (cast<int32> ?#4))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        9,
        3,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 64))",
    ),
    (
        11,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))",
    ),
    (
        12,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))",
    ),
    (
        14,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 128))",
    ),
    (
        15,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 256))",
    ),
    (
        47,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        48,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        49,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        50,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        51,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        52,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        53,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        58,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?#6) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        59,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        60,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        161,
        3,
        "(- (cast<uint8x128> ?x0) (cast<int32x128> (call[widen_right_mul] (cast<int32x128> ?x1) (cast<int16x128> (call[x128] (cast<int16> (call[int16] (cast<uint8> ?#2))))))))",
    ),
    (
        181,
        3,
        "(* (cast<int32> ?x0) (cast<int32> 256))",
    ),
    (
        197,
        3,
        "(== (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        200,
        3,
        "(if (cast<int32x128> (get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<bool> ?x1))",
    ),
    (
        211,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        214,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        230,
        3,
        "(get ?#1 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> 1) (cast<int32> 64))))",
    ),
    (
        24,
        4,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> 0) (cast<int32> ?x1))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> 0) (cast<int32> (+ (cast<int32> ?x3) (cast<int32> (neg (cast<int32> 1))))))) (cast<int32> ?x4))) (cast<int32> ?x5))))",
    ),
    (
        41,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        154,
        4,
        "(- (cast<int32> (call[max] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> 0) (cast<int32> ?x3))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x4) (cast<int32> 0) (cast<int32> (+ (cast<int32> ?x5) (cast<int32> (neg (cast<int32> 1))))))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x8))) (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x9) (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> ?x3) (cast<int32> 0))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x4) (cast<int32> ?x5) (cast<int32> 1))) (cast<int32> ?x6))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x7))))) (cast<int32> ?x10))))",
    ),
    (
        163,
        4,
        "(* (cast<int32> (+ (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> 1))) (cast<int32> ?x2))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x3))",
    ),
    (
        180,
        4,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        192,
        4,
        "(<= (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        19,
        6,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?#5) (cast<int32> ?x2))) (cast<int32> (* (cast<int32> ?x3) (cast<int32> ?x4))))))",
    ),
    (
        37,
        6,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        42,
        6,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 2))))",
    ),
    (
        152,
        6,
        "(- (cast<int32> (* (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> ?x1))",
    ),
    (
        174,
        6,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x1))",
    ),
    (
        177,
        6,
        "(* (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 0))) (cast<int32> 2))",
    ),
    (
        178,
        6,
        "(* (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> 0))) (cast<int32> 2))",
    ),
    (
        21,
        7,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        151,
        7,
        "(- (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        198,
        7,
        "(&& (cast<bool> ?x0) (cast<bool> ?x1))",
    ),
    (
        74,
        8,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        176,
        8,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        68,
        9,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> 1))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        80,
        9,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 384) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> 1))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        194,
        9,
        "(<= ?x0 ?x1)",
    ),
    (
        175,
        10,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        196,
        10,
        "(== (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        188,
        12,
        "(< (cast<int32> ?x0) (cast<int32> 0))",
    ),
    (
        207,
        12,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        8,
        19,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        39,
        19,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        182,
        20,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        45,
        42,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        160,
        43,
        "(- (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
]
